### PR TITLE
Sync writes/deletes on sub case search indices.

### PR DESCRIPTION
**Pull request by AmitPhulera**
## Product Description

This PR builds up on https://github.com/dimagi/commcare-hq/pull/34663 and adds ability to sync writes and deletes to sub-case search indices. 
It uses a django-setting called `CASE_SEARCH_SUB_INDICES` which will look something like 
```
{
    'co-carecoordination-perf': {
        'index_cname': 'case_search_bha',
        'multiplex_writes': True,
    }
}
```

This setting will be used to decide which sub-index should the duplicate writes be routed to.

This PR in some ways combine commits from https://github.com/dimagi/commcare-hq/pull/34602 with certain modifications to make the changes less error prone.

Review by commit 🐡 

## Feature Flag

NA
## Safety Assurance
Most of the changes(except pillow delete) have been covered by tests and have been tested locally by running case-pillow, submitting forms and ensuring that data is present in both the indices. Then archiving the form to ensure that the case is deleted from both case-search indices.

Another thing to note here is that the changes are added in a way that if setting is not enabled on the domain then it should not have any effect. I am also putting this on staging to test out the changes with `co-carecoordination-test` on staging.

### Automated test coverage

Tests are added for most of the changes.
### QA Plan
Given the nature of the change it would be a good idea to get QA done but given time sensitivity of the work, I am not sure if we should do it.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/es/client.py
Comment (millerdev) [inline corehq/apps/es/client.py:1159]: This function seems to be at the wrong abstraction level. `ElasticMultiplexAdapter` should not need to know about any particular index by name except for the ones it's configured to multiplex via `primary_adapter`/`secondary_adapter`.

File: corehq/ex-submodules/pillowtop/processors/elastic.py
Comment (millerdev) [inline corehq/ex-submodules/pillowtop/processors/elastic.py:75]: dot-syntax is equivalent and preferred to using `getattr`
```suggestion
                meta = change.metadata
```

Comment (AmitPhulera) [reply corehq/ex-submodules/pillowtop/processors/elastic.py:75]: This is to ensure that if 'metadata' is None or not set it will not error out.

File: corehq/apps/es/tests/test_case_search_adapter.py
Comment (gherceg) [inline corehq/apps/es/tests/test_case_search_adapter.py:121]: Could this live in the setUp method instead of here? I think it just gets in the way of understanding what the test is testing.

File: corehq/apps/es/tests/test_case_search_adapter.py
Comment (gherceg) [inline corehq/apps/es/tests/test_case_search_adapter.py:126]: I think `self.assertCountEqual` is better to prevent a potential flaky test if results are returned in a different order between the two indices?

File: corehq/ex-submodules/pillowtop/processors/elastic.py
Comment (gherceg) [reply corehq/ex-submodules/pillowtop/processors/elastic.py:75]: To prevent an error, we would need to specify a default option in `getattr` in case the attribute does not exist.

**Review by gherceg — COMMENTED** [review]
First pass review. I can make the minor changes needed so that this PR is ready for Monday.

File: corehq/ex-submodules/pillowtop/processors/elastic.py
Comment (gherceg) [reply corehq/ex-submodules/pillowtop/processors/elastic.py]: Added default value in 2a81942dc96dd45d7db41f309382554f0520ff25 to prevent an error from being raised, which is the reason to use getattr here.

File: corehq/apps/es/tests/test_case_search_adapter.py
Comment (gherceg) [reply corehq/apps/es/tests/test_case_search_adapter.py:121]: This isn't a blocking comment and I don't plan to make any changes to the current PR related to this, just wanted to flag it.

File: corehq/apps/es/tests/test_case_search_adapter.py
Comment (gherceg) [reply corehq/apps/es/tests/test_case_search_adapter.py:126]: This isn't a blocking comment and I don't plan to make any changes to the current PR related to this, just wanted to flag it.

File: corehq/apps/es/case_search.py
Comment (esoergel) [inline corehq/apps/es/case_search.py:194]: This is the secondary adapter right?  I'd use some language in here to clarify, since `self` is also an adapter

File: corehq/ex-submodules/pillowtop/processors/elastic.py
Comment (esoergel) [inline corehq/ex-submodules/pillowtop/processors/elastic.py]: Is there a chance this won't be present either?  I think you said that happens when deleting a full domain, is that right?

File: corehq/apps/es/case_search.py
Comment (esoergel) [reply corehq/apps/es/case_search.py:194]: I see elsewhere you called it `sub_index_adapter`, that seems nice and clear - think it'd be appropriate to add it here too?

File: corehq/ex-submodules/pillowtop/processors/elastic.py
Comment (esoergel) [reply corehq/ex-submodules/pillowtop/processors/elastic.py]: I see there's conversation about this elsewhere - will move to that thread.

File: corehq/ex-submodules/pillowtop/processors/elastic.py
Comment (esoergel) [reply corehq/ex-submodules/pillowtop/processors/elastic.py]: I do still feel a bit uncomfortable with the prospect of `domain` maybe being missing - that means when that situation arises, we won't multiplex.

Possible alternatives that come to mind are:
1. We identify specific circumstances where that occurs, and make available an explicit signal that it's okay to not multiplex in this instance (assuming it _is_ okay)
2. Since this is only for deletions, could we propagate the deletion to _all_ indices?  Assuming it noops if the doc isn't present, that seems like it'd be in line with our integrity model.

We may also want to log an exception even if we don't fail hard.

File: corehq/apps/es/case_search.py
Comment (gherceg) [reply corehq/apps/es/case_search.py]: 02b8735d9eba66f3849ae96e74ab5807ae959971

File: corehq/ex-submodules/pillowtop/processors/elastic.py
Comment (gherceg) [reply corehq/ex-submodules/pillowtop/processors/elastic.py]: I added f84af4aa1c012a826c67bba39c7ce23fce14e20d to at least log an error if we attempt to delete a case from case search without specifying a domain. Based on a slack conversation with Ethan, it seems pretty unlikely that we will find ourselves in a scenario where the domain is None. Currently on production, there aren't any cases with a domain set to None, which means any soft deletion will have a domain available since the postgres doc is still available to query. For hard deletions where the postgres doc is no longer available, we can be fairly confident that the ChangeMeta will be populated with the domain based on [publish_case_deleted](https://github.com/dimagi/commcare-hq/blob/2a81942dc96dd45d7db41f309382554f0520ff25/corehq/form_processor/change_publishers.py#L56-L64). This is because our hard deletion logic only deletes a case if the provided case_id **and** domain match, so if the caller didn't specify a domain, the case deletion would fail, and as stated previously, we don't currently have any cases where a domain is None.

**Review by gherceg — APPROVED** [review]
Approving with the rationale that if any issues around writing to the BHA index come up after merging, that is still low risk as we can stop reading from that index by updating a toggle.

(My approval is worthless to github since I made changes here too)

File: settings.py
Comment (calellowitz) [inline settings.py:1082]: nit: rather than give an example this would be more helpful if it explained what the fields do/mean. for example, i think the first is a domain, but only because I recognize that domain, and I am not sure what the other two fields control looking here.

File: corehq/ex-submodules/pillowtop/processors/elastic.py
Comment (calellowitz) [inline corehq/ex-submodules/pillowtop/processors/elastic.py:108]: why does this not need to check the change meta like the above did?

File: settings.py
Comment (calellowitz) [inline settings.py:1083]: where is the cname mapped to an actual index name?

**Review by calellowitz — APPROVED** [review]
A couple nits and questions and I am not super familiar with the adapter methods, but this all looks reasonable

File: corehq/ex-submodules/pillowtop/processors/elastic.py
Comment (gherceg) [reply corehq/ex-submodules/pillowtop/processors/elastic.py:108]: I think it is because this isn't a deletion, and therefore there should still be a valid DB object when we call `change.get_document()`.

File: settings.py
Comment (gherceg) [reply settings.py:1083]: The [case_search_bha_adapter](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/es/case_search_bha.py#L36-L41) effectively maps cname to the actual index name. So using the cname defined in settings, we can grab the adapter as done here: https://github.com/dimagi/commcare-hq/pull/34673/files#diff-7ba2ac2cd9c42581b497ab0c7cb5ea77288d3a8d7e8583cfc2c35a4da9e063b6R232

File: corehq/apps/es/client.py
Comment (snopoke) [reply corehq/apps/es/client.py:1159]: This also seemed a bit odd to me. Looking at the implementation again it seems like the multiplex adapter doesn't actually call the indexing methods on the primary or secondary but instead uses it's own logic which is why changes need to be applied here as well.

But I do agree with @millerdev that it would be ideal if the multiplex adapter didn't have to know about this. One option could be to have `_render_bulk_action` return a list of actions instead of a single one. That would allow the primary adapter to generate the actions for both the main and sub index.

(I'm OK with the current implementation for the short term)

**Comment by gherceg** [comment]
Acknowledging that we will followup on this to find a better long term solution, but we want to get a first implementation of multiplexing merged so that we can test using the new index. For that reason, I'm going to merge this PR.

**Pull request merged**
